### PR TITLE
Make dockershim periodically clean up its `containerCleanupInfos`

### LINKED ISSUE #74843: Make dockershim periodically clean up its `containerCleanupInfos` map
**What would you like to be added**:
Periodically remove cleanupInfos for old containers from dockershim's `containerCleanupInfos` map.

One way of doing that would be to just forget about `cleanupInfo`s older than say an hour, as they should always be short-lived.

**Why is this needed**:
https://github.com/kubernetes/kubernetes/pull/74737 introduced a new in-memory map for the dockershim, that could potentially (in pathological cases) cause memory leaks - for containers that use GMSA cred specs, get created successfully, but then never get started nor removed.

This might be okay for the alpha release of the GMSA feature, but should be addressed in the beta.

See https://github.com/kubernetes/kubernetes/pull/74737#discussion_r261784181

### PR DESCRIPTION
**Pull request by wk8**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

https://github.com/kubernetes/kubernetes/pull/74737 introduced a new in-memory
map for the dockershim, that could potentially (in pathological cases) cause
memory leaks - for containers that use GMSA cred specs, get created
successfully, but then never get started nor removed.

This patch adds a new goroutine in the dockershim that periodically cleans
up stale `containerCleanupInfos`.

**Which issue(s) this PR fixes**:
Fixes #74843

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by neolit123** [comment]
@kubernetes/sig-node-pr-reviews 
/priority important-longterm

**Comment by mattjmcnaughton** [comment]
@wk8 this looks to be the same as https://github.com/kubernetes/kubernetes/pull/77953. Should I hold on reviewing this until you've closed your your WIP pr? Do you mind also marking this one as WIP in the meantime?

Thanks :)

**Comment by dims** [comment]
@cblecker @thockin this has spanking brand new dependency!! `github.com/wk8/go-ordered-map`

**Comment by dims** [comment]
/hold

(hold for @kubernetes/dep-approvers to chime in)

**Comment by cblecker** [comment]
Why are we importing a new personal library from your personal repo, @wk8? I'm not comfortable with this, and it seems unnecessary.

**Comment by dims** [comment]
let's move this out please. it's clearly not ready because of the dependency issues at least.

/milestone v1.16

**Comment by wk8** [comment]
@cblecker : an ordered dict seems to be the natural data structure to efficiently do what we need here. I didn't find one in the existing dependencies, hence why I added this new one.

If you feel strongly against adding it, we can either:
1. iterate over all existing cleanup infos at each tick, which should be fine in most cases, as there are not supposed to ever be too many of them - but then this whole change is precisely supposed to clean up in case of pathological dysfunctions, so I didn't want to make the assumption that iterating every time was fine
2. the library's code itself is only 100 lines, we could just add it to this repo?

Please let me know what you think.

**Comment by wk8** [comment]
@dims : agreed not ready, I need to fix failing tests. Put the PR out mostly to get feedback on the general shape of the change - including the new dep - before spending too much time on it.

File: pkg/kubelet/dockershim/docker_container_cleanup_manager.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+	cm.unsafePerformCleanup(containerID)
+}
+
+// unsafePerformCleanup is the same as performCleanup, but assumes the manager's lock has already
+// been acquired.
+func (cm *containerCleanupManager) unsafePerformCleanup(containerID string) {
+	if cleanupInfoWithTimestamp, present := cm.cleanupInfos.Delete(containerID); present {
+		cleanupInfo := cleanupInfoWithTimestamp.(*containerCleanupInfoWithTimestamp).cleanupInfo
+		cm.cleanupService.performContainerCleanup(containerID, cleanupInfo)
```
Comment (ddebroy) [inline pkg/kubelet/dockershim/docker_container_cleanup_manager.go]: Should `performContainerCleanup` bubble up the error from `performPlatformSpecificContainerCleanup` here so we can check for errors and insert back `<containerID, cleanupInfo>` to `cleanupInfos` for retrying the cleanup in future?

**Review by ddebroy — COMMENTED** [review]
May be the ordered map implementation (with unit tests) can be moved within `pkg/util` to avoid a fresh vendoring and the concerns around that.

Also have a comment below about retrying failed cleanups of the persistent registry data.

**Comment by cblecker** [comment]
@wk8 Are you reviewing these failures before retesting? Some of them are legitimate. Please don't spam the CI without reviewing test failures.

**Comment by cblecker** [comment]
@wk8 These are legitimate failures. Please stop spamming CI.

**Comment by cblecker** [comment]
@wk8 You've been repeatedly asked to stop spamming CI for legitimate failures. I'm closing this PR at this time. Please @ me when you have fixed the failures and tested locally.

**Comment by ddebroy** [comment]
Looks like the go-lint failures indeed need fixing:
```
pkg/util/orderedmap/orderedmap.go:30:6: exported type Pair should have comment or be unexported
pkg/util/orderedmap/orderedmap.go:37:6: exported type OrderedMap should have comment or be unexported
pkg/util/orderedmap/orderedmap.go:114:1: comment on exported method Pair.Prev should be of the form "Prev ..."
```

**Comment by BenTheElder** [comment]
Other than golint all of the e2e tests seem to fail
https://prow.k8s.io/pr-history/?org=kubernetes&repo=kubernetes&pr=77886

Would need to look into the logs to see why, but it's easily consistent
enough to be something in this PR.

**Comment by wk8** [comment]
Re-opened at https://github.com/kubernetes/kubernetes/pull/80320

**Comment by cblecker** [comment]
@wk8 Why were the comments going back and being deleted? There were 41 retests in a 3 hour span.

Please do not run these types of scripts against Kubernetes PRs. They are unable to actually interpret test results.

**Comment by wk8** [comment]
@cblecker : apologies again, I understand this was not ideal.

Sadly there are enough flakes that auto-retrying tests does come in handy quite often on k8s PRs (as evidenced by eg https://github.com/kubernetes/kubernetes/pull/80320, where out of 6 initially failing checks, 3 have now passed without any changes).

I have limited auto-retries to 2 to avoid that kind of spam going forward. Sorry again.

**Pull request closed without merging**